### PR TITLE
Configurable promotion adjuster

### PR DESCRIPTION
**Pull request by mamhoff**
## Summary

This adds an extension point for a configurable promotion adjuster to the app configuration, and uses that extension point to point to a new  `Spree::Promotion::LegacyAdjuster`. This LegacyAdjuster does all the things that the order updater in conjunction with the Spree::Adjustment class was doing. 

## Checklist

Check out our [PR guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines) for more details.

The following are mandatory for all PRs:

- [x] I have written a thorough PR description.
- [x] I have kept my commits small and atomic.
- [x] I have used clear, explanatory commit messages.

The following are not always needed (~cross them out~ if they are not):

- [x] I have added automated tests to cover my changes.
- [ ] ~I have attached screenshots to demo visual changes.~
- [ ] ~I have opened a PR to update the [guides](https://github.com/solidusio/edgeguides).~
- [ ] ~I have updated the readme to account for my changes.~

### CODE REVIEW

File: core/app/models/spree/promotion/legacy_adjuster.rb
Comment (kennyadsl) [inline core/app/models/spree/promotion/legacy_adjuster.rb:7]: Not mandatory, but a more detailed (high-level) explanation could help readers in understanding what's going on here without reading the whole class code.

File: core/app/models/spree/promotion/legacy_adjuster.rb
Comment (kennyadsl) [inline core/app/models/spree/promotion/legacy_adjuster.rb:8]: If this is the only solution provided in core, I don't think we should call it "Legacy". Or am I missing something?

File: core/lib/spree/app_configuration.rb
Comment (kennyadsl) [inline core/lib/spree/app_configuration.rb:345]: This comment needs to be changed to `promotion_adjuster_class`

**Review by kennyadsl — CHANGES_REQUESTED** [review]
Love it! I left a couple of things I think should be changed, though. 

Also, can you please provide a real-world example of when someone could use a custom version of this? Not because I don't think there will be, but it will be super useful when writing a how-to for this preference based on a real need.

Thanks Martin! 🙏

File: core/app/models/spree/promotion/legacy_adjuster.rb
Comment (mamhoff) [reply core/app/models/spree/promotion/legacy_adjuster.rb:8]: The plan is another adjuster that does not call `Spree::Adjustment#recalculate`, as that code path is very difficult to optimize and quite error-prone in some situations. This one does call that method, which I think should be deprecated once we have an alternative.

Comment (mamhoff) [reply core/app/models/spree/promotion/legacy_adjuster.rb:12]: The adjustment should not be recalculated using its source; I'd much rather load all promotions and determine their eligibilty first, and then add, remove or change adjustments. The way it is done right now falls apart when using multi-level (orde,r line item, shipment, shipping rate) rules.

Comment (kennyadsl) [reply core/app/models/spree/promotion/legacy_adjuster.rb:12]: I agree it's not the best way to handle the recalculation, but we risk that this legacy class ends up in a new version without the new option. My proposal is:

- rename it into something that works no matter if there's the new option available. Some semantic name that describes what it does would be awesome (and I know, hard to find as hell 😄 ). 
- merge it
- if we merge the new option before the next minor/major, we are still allowed to rename this class without deprecating it, because it has never been released
- if we don't merge the new option before the next minor/major, we can still deprecate the class, even if doesn't have the legacy suffix in its name. 

What do you think?

**Review by adammathys — APPROVED** [review]

File: core/spec/lib/spree/app_configuration_spec.rb
Comment (kennyadsl) [inline core/spec/lib/spree/app_configuration_spec.rb:35]: Last occurrence of legacy here

**Review by kennyadsl — APPROVED** [review]
Approved but if you have time to fix that occurrence of legacy it would be great!

File: core/spec/lib/spree/app_configuration_spec.rb
Comment (mamhoff) [reply core/spec/lib/spree/app_configuration_spec.rb]: fixed

**Review by jarednorman — APPROVED** [review]

File: core/app/models/spree/order_updater.rb
Comment (kennyadsl) [inline core/app/models/spree/order_updater.rb:165]: @mamhoff while taking a look at this PR again, I notice this line. Is this comment still valid? I think we are already doing this in the `update_promotion` method so we can probably already remove this line?

Comment (kennyadsl) [reply core/app/models/spree/order_updater.rb:165]: I diwscussed with @mamhoff about this, and it is actually valid. People might be calling `update_item_promotions` in their own store, which doesn't include this `promo_total` calculation.

Comment (mamhoff) [reply core/app/models/spree/order_updater.rb:165]: https://github.com/solidusio/solidus/pull/4890

**Pull request merged**
